src/pybind/mgr/dashboard/controllers/role.py

```diff
@@ -1,15 +1,110 @@
 # -*- coding: utf-8 -*-
 from __future__ import absolute_import
 
-from . import ApiController, RESTController
-from ..security import Scope
+import cherrypy
+
+from . import ApiController, RESTController, UiApiController
+from ..exceptions import RoleDoesNotExist, DashboardException,\
+    RoleIsAssociatedWithUser, RoleAlreadyExists
+from ..security import Scope as SecurityScope, Permission
 from ..services.access_control import ACCESS_CTRL_DB, SYSTEM_ROLES
 
 
-@ApiController('/role', Scope.USER)
+@ApiController('/role', SecurityScope.USER)
 class Role(RESTController):
+    @staticmethod
+    def _role_to_dict(role):
+        role_dict = role.to_dict()
+        role_dict['system'] = role_dict['name'] in SYSTEM_ROLES
+        return role_dict
+
+    @staticmethod
+    def _validate_permissions(scopes_permissions):
+        if scopes_permissions:
+            for scope in scopes_permissions.keys():
+                if scope not in SecurityScope.all_scopes():
+                    raise DashboardException(msg='Invalid scope',
+                                             code='invalid_scope',
+                                             component='role')
+                permissions = scopes_permissions[scope]
+                if any(permission not in Permission.all_permissions()
+                       for permission in permissions):
+                    raise DashboardException(msg='Invalid permission',
+                                             code='invalid_permission',
+                                             component='role')
+
+    @staticmethod
+    def _set_permissions(role, scopes_permissions):
+        role.reset_scope_permissions()
+        if scopes_permissions:
+            for scope in scopes_permissions.keys():
+                permissions = scopes_permissions[scope]
+                if permissions:
+                    role.set_scope_permissions(scope, permissions)
+
+    def list(self):
+        roles = dict(ACCESS_CTRL_DB.roles)
+        roles.update(SYSTEM_ROLES)
+        roles_items = sorted(roles.items(), key=lambda role: role[1].name)
```

Comment:
maybe use the dict values directly like: `roles = sorted(roles.values(), key=lambda role: role.name)`
and change the following line to: `[Role._role_to_dict(r) for r in roles]`